go/vt/vtgate/grpcvtgateservice/server.go

```diff
@@ -95,8 +97,10 @@ func immediateCallerIDFromCert(ctx context.Context) (string, []string) {
 }
 
 func immediateCallerID(ctx context.Context) (string, []string) {
-	if immediate := servenv.StaticAuthUsernameFromContext(ctx); immediate != "" {
-		return immediate, nil
+	if useStaticAuthenticationIdentity {
```

Comment:
It seems like this check is in the wrong place? Isn't it more a responsibility of the caller of `immediateCallerID`? 

We already have the `useEffective` so shouldn't we make sure that if that value is set, we don't let immediate override it?

Reply:
I think if we go with an approach like the following, we can achieve the same result without needing another flag?

```patch
diff --git i/go/vt/vtgate/grpcvtgateservice/server.go w/go/vt/vtgate/grpcvtgateservice/server.go
index d012786d6e..a819730cb1 100644
--- i/go/vt/vtgate/grpcvtgateservice/server.go
+++ w/go/vt/vtgate/grpcvtgateservice/server.go
@@ -104,12 +104,15 @@ func immediateCallerID(ctx context.Context) (string, []string) {
 // withCallerIDContext creates a context that extracts what we need
 // from the incoming call and can be forwarded for use when talking to vttablet.
 func withCallerIDContext(ctx context.Context, effectiveCallerID *vtrpcpb.CallerID) context.Context {
-       immediate, securityGroups := immediateCallerID(ctx)
-       if immediate == "" && useEffective && effectiveCallerID != nil {
+       var immediate string
+       var securityGroups []string
+       if useEffective && effectiveCallerID != nil {
                immediate = effectiveCallerID.Principal
                if useEffectiveGroups && len(effectiveCallerID.Groups) > 0 {
                        securityGroups = effectiveCallerID.Groups
                }
+       } else {
+               immediate, securityGroups = immediateCallerID(ctx)
        }
        if immediate == "" {
                immediate = unsecureClient
```

Reply:
The changes @dbussink is suggesting above make sense to me. If the intention is for effective caller id to take precedence over other credentials, then that could be done explicitly here. Just reverting the changes in #12050 would mean that a client connecting to vtgate using mTLS would still have their immediate caller id set from the client cert rather than the effective caller id.

Reply:
@brendar in your use case do you use mTLS and not want to use the client name from the cert?

Reply:
No, we use TLS (but not mTLS) + static auth to connect to vtgates. The precedence order doesn't matter to us since we're also not using `grpc_use_effective_callerid`, but we'd like to be able to use grpc static auth usernames in table ACLs. That allows us to provide clients with a set of username/password credentials, and they can use those to connect to vtgate via mysql protocol or grpc as they choose.

To clarify what I mean by precedence order, before #12050 the immediate caller id would have been set from the first non-empty value from:
* The client cert common name (if using mTLS)
* The effective caller id (if `--grpc_use_effective_callerid=true`)

After #12050 the order would have been:
* The static auth username (if using static auth)
* The client cert common name (if using mTLS)
* The effective caller id (if `--grpc_use_effective_callerid=true`)

If the effective caller id should take precedence over other credentials, then perhaps the order should be something like this?
* The effective caller id (if `--grpc_use_effective_callerid=true`)
* The static auth username (if using static auth)
* The client cert common name (if using mTLS)

Looking at the [comment for grpc_use_effective_callerid](https://github.com/vitessio/vitess/blob/d6bc90baef4013a99ea4e346b4bcfa2300090d2b/go/vt/vtgate/grpcvtgateservice/server.go#L54) though
> "If set, and SSL is not used, will set the immediate caller id from the effective caller id's principal."

If we wanted to preserve that behavior, then perhaps the order should be:
* The client cert common name (if using mTLS)
* The effective caller id (if `--grpc_use_effective_callerid=true`)
* The static auth username (if using static auth)